ships/0036-runAs-for-supporting-steps.md

```diff
@@ -0,0 +1,126 @@
+
+#### Story 2
+
+As a Shipwright Build Strategy author, I want to read about recommendations to specify runAsUser and runAsGroup so that I correctly define the steps.
+
+## Proposal
+
+### Base image changes
+
+Container images define which users and groups exist in the container by specifying the content of `/etc/passwd` and `/etc/group`. Though, it is generally possible to run a container with as a user or group that is not included in the container image. We are making use of this.
+
+Unfortunately, binaries may rely on an entry in those files to be present to properly work. For the binaries that we run in supporting steps, this is the case for `ssh` which we need in the Git step when cloning using the SSH protocol. [SSH requires an entry and fails if it is not present.](https://github.com/openssh/openssh-portable/blob/V_9_3_P1/ssh.c#L671-L676)
+
+To handle this, we will change the base image for the Git step to contain empty `/etc/passwd` and `/etc/group` files that are writable for everybody. The Git step will instead register itself before running and `ssh` commands.
+
+To mitigate the security problem that this opens (the user can exec into that container, write to `/etc/passwd` and `su` to become root / user 0), we will change the default configuration of our supporting steps to not allow privilege escalation and to drop all capabilities. We actually should have configured them like this all the time already to give them no unnecessary privileges.
+
+### BuildRun reconciler changes
+
+The BuildRun reconciler is changed to check the steps of the build strategy. If all of the included steps define a runAsUser and runAsGroup that is always the same, then the Shipwright-added steps will be changed to run as this user and/or group.
+
+Otherwise, the runAsUser and runAsGroup that are defined in the configuration's container templates are used.
+
+### Sample updates
+
+The prepare step of all build strategies is removed. The Paketo build strategy is updated to use the Jammy builder and run as user 1001.
+
+### Documentation updates
+
+The build strategy documentation is updated to explain the consequences of defining, or not defining runAsUser and runAsGroup on the steps.
+
+### Test Plan
+
+Unit and integration tests should cover the part that ensures that everything behaves correctly. Our existing e2e tests should be able to cover the changes to the build strategies.
+
+The e2e tests that depend on a Git secret have to succeed locally as they do not run as part of a build.
+
+## Release Criteria
+
+## Risks and Mitigations
+
+- Risk: we might break existing build strategies that run steps as a non-root non-1000 user and were able to work with source code cloned as 1000 without the prepare step hack. I don't think that this is realistically the case. But, it can't be ruled out completely. The release notes and release blog post should highlight the change and link to the build strategy documentation change.
```

Comment:
Yeah, I understand your concern. Let's brainstorm on Monday. Based on today's build strategies I basically see two patterns:

* Strategies like buildpacks or buildkit or ko run with a single step as 1000/1000 and this works fine because our source step suddenly also runs as 1000/1000. The prepare step we have there these days is actually only needed for cases where one changes shipwright's configuration to run the source steps as a different user. And no, we were not thinking about that before, but simply have those steps for historic reasons back from times when the git clone was performed by Tekton's resource as root.
* Strategies like kaniko or buildah that run as root (buildah even privileged) and do not need to care about who owns the source files because they can do anything with them because they are root anyway.

The new code btw would run all those strategies just fine without that you need to change them.